_source/_posts/2022-07-21-jhipster-ci-cd.md

```diff
@@ -0,0 +1,865 @@
+
+### Trigger the CI-CD pipeline with a GitHub push
+
+For testing the workflow, make a code change in the gateway. Edit `src/main/webapp/content/scss/_bootstrap-variables.scss` and update the following variable:
+
+```scss
+$body-bg: steelblue;
+```
+Also, the pipeline under test will only trigger if a new image tag is detected. So edit `.circleci/confg.yml` and update the image name:
+
+```yml
+IMAGE_NAME: indiepopart/gateway:v1
+```
+
+Commit and push the change to the `main` branch, and watch the CircleCI CI pipeline triggers. After the new gateway image is pushed to DockerHub, watch the Spinnaker CD pipeline trigger and deploy the updated gateway. On the left menu choose **Clusters**, and verify the active gateway deployment now has a _V002_. If you click over the **V002** box, you can also verify the image tag that was deployed.
+
+{% img blog/jhipster-ci-cd/sp-version-2.png alt:"Spinnaker deployment version 2" width:"800" %}{: .center-image }
+
+### Inspect Spinnaker logs
+
+For pipeline and trigger debugging, the Spinnaker services `spin-echo` and `spin-igor` inform docker monitoring events and also the reasons why an execution was skipped.
+
+```shell
+kubectl get pods -n spinnaker
+kubectl logs spin-igor-7c8bdd94f5-lx5dl -n spinnaker | grep v1
+```
+```text
+2022-08-05 03:20:27.285  INFO 1 --- [   scheduling-1] c.n.spinnaker.igor.docker.DockerMonitor  : Found 1 new images for docker-account. Images: [{imageId=igor:dockerRegistry:v2:docker-account:indiepopart/gateway:v1, sendEvent=true}]
+2022-08-05 03:20:27.286  INFO 1 --- [   scheduling-1] c.n.spinnaker.igor.docker.DockerMonitor  : New tagged image: account=docker-account, image=igor:dockerRegistry:v2:docker-account:indiepopart/gateway:v1. Digest is now [null].
+2022-08-05 03:20:27.288  INFO 1 --- [   scheduling-1] c.n.spinnaker.igor.docker.DockerMonitor  : Sending tagged image info to echo: account=docker-account: image=igor:dockerRegistry:v2:docker-account:indiepopart/gateway:v1
+```
+
+## Know about Spinnaker's best practices and features
+
+When implementing continuous delivery, be aware of some best practices and features provided by Spinnaker described below.
+
+- **Deploy Docker images by digest** Spinnaker documentation recommends deploying images by digest instead of tag, because the tag might reference different content each time. The digest is a content-based hash of the image and it uniquely identifies it. Spinnaker's artifact substitution allows deploying the same manifest, with the image digest supplied by the pipeline trigger. Using the proposed trigger and a free DockerHub account as the registry, the digest data seems not to be available. It is also recommended to trigger a CD pipeline off of a push to a docker registry instead of a GitHub push or Jenkins job, allowing development teams to choose their delivery process without more constraints.
```

Comment:
```suggestion
- **Deploy Docker images by digest**: Spinnaker documentation recommends deploying images by digest instead of tag, because the tag might reference different content each time. The digest is a content-based hash of the image and it uniquely identifies it. Spinnaker's artifact substitution allows deploying the same manifest, with the image digest supplied by the pipeline trigger. Using the proposed trigger and a free DockerHub account as the registry, the digest data seems not to be available. It is also recommended to trigger a CD pipeline off of a push to a Docker registry instead of a GitHub push or Jenkins job, allowing development teams to choose their delivery process without more constraints.
```
capitalizing "Docker" consistently, unless that's not correct...